lib/active_merchant/billing/gateways/authorize_net.rb

```diff
@@ -481,30 +675,92 @@ def message_from(response, avs_result, cvv_result)
           end
         end
 
-        response[:response_reason_text]
+        response[:response_reason_text] || response[:message_text]
       end
 
-      def authorization_from(response)
-        [response[:transaction_id], response[:account_number]].join("#")
+      def authorization_from(action, response)
+        if action == :cim_store
+          [response[:customer_profile_id], response[:customer_payment_profile_id], action].join("#")
+        else
+          [response[:transaction_id], response[:account_number], action].join("#")
+        end
       end
 
       def split_authorization(authorization)
-        transaction_id, card_number = authorization.split("#")
-        [transaction_id, card_number]
+        authorization.split("#")
```

Comment:
Shouldn't break 'em since for the existing behavior we're only accessing the first two.  And if the last element is empty, we use the current behavior.